Library/OcAppleImg4Lib/OcAppleImg4Lib.c

```diff
@@ -0,0 +1,316 @@
+  return VerifySignatureFromData (
+           Modulus,
+           ModulusSize,
+           Exponent,
+           Signature,
+           SignatureSize,
+           Data,
+           DataSize,
+           AlgoType
+           );
+}
+
+VOID
+InternalRetrieveHwInfo (
+  OUT DERImg4Environment  *Environment
+  )
+{
+  EFI_STATUS Status;
+  UINTN      DataSize;
+  //
+  // TODO: We probably want an internal structure to collect and store these.
+  //
+  ASSERT (Environment != NULL);
+
+  ZeroMem (Environment, sizeof (*Environment));
+
+  DataSize = sizeof (Environment->ecid);
+  Status = gRT->GetVariable (
+                  L"ApECID",
+                  &gAppleSecureBootVariableGuid,
+                  NULL,
+                  &DataSize,
+                  &Environment->ecid
+                  );
+  if (EFI_ERROR (Status) || DataSize != sizeof (Environment->ecid)) {
+    Environment->ecid = 0;
+  }
+
+  DataSize = sizeof (Environment->chipId);
+  Status = gRT->GetVariable (
+                  L"ApChipID",
+                  &gAppleSecureBootVariableGuid,
+                  NULL,
+                  &DataSize,
+                  &Environment->chipId
+                  );
+  if (EFI_ERROR (Status) || DataSize != sizeof (Environment->chipId)) {
+    Environment->chipId = 0;
+  }
+
+  DataSize = sizeof (Environment->boardId);
+  Status = gRT->GetVariable (
+                  L"ApBoardID",
+                  &gAppleSecureBootVariableGuid,
+                  NULL,
+                  &DataSize,
+                  &Environment->boardId
+                  );
+  if (EFI_ERROR (Status) || DataSize != sizeof (Environment->boardId)) {
+    Environment->boardId = 0;
+  }
+
+  Environment->certificateEpoch = 2;
+
+  DataSize = sizeof (Environment->securityDomain);
+  Status = gRT->GetVariable (
+                  L"ApSecurityDomain",
+                  &gAppleSecureBootVariableGuid,
+                  NULL,
+                  &DataSize,
+                  &Environment->securityDomain
+                  );
+  if (EFI_ERROR (Status) || DataSize != sizeof (Environment->securityDomain)) {
+    Environment->securityDomain = 1;
+  }
+
+  DataSize = sizeof (Environment->productionStatus);
+  Status = gRT->GetVariable (
+                  L"ApProductionStatus",
+                  &gAppleSecureBootVariableGuid,
+                  NULL,
+                  &DataSize,
+                  &Environment->productionStatus
+                  );
+  if (EFI_ERROR (Status)
+   || DataSize != sizeof (Environment->productionStatus)) {
+    Environment->productionStatus = TRUE;
+  }
+
+  DataSize = sizeof (Environment->securityMode);
+  Status = gRT->GetVariable (
+                  L"ApSecurityMode",
+                  &gAppleSecureBootVariableGuid,
+                  NULL,
+                  &DataSize,
+                  &Environment->securityMode
+                  );
+  if (EFI_ERROR (Status) || DataSize != sizeof (Environment->securityMode)) {
+    Environment->securityMode = TRUE;
```

Comment:
It could be added for self-documentation reasons, but the default is TRUE, so if it's any value but TRUE, it'll yield TRUE in all boolean usages anyway. Maybe a comment?

Reply:
I do not like this, as BOOLEAN is not necessarily _Bool in EDK II. You can do & 1 with real bool, but with EDK II you cannot. Thus I believe it is a must that we ensure ourselves that BOOLEAN only has 0 and 1 values, or it is not BOOLEAN.

Reply:
Fair point, I didn't think of that operator effect. But I also think there is a much more important question: https://github.com/acidanthera/OcSupportPkg/blob/8cba450bfd99bdf659e43c1cb5ba4f07f0a45c55/Library/OcAppleImg4Lib/OcAppleImg4Lib.c#L86
Will we be consumer (user sets them e.g. via NVRAM section) or exposer (user sets them maybe as part of extended SMBIOS with a simplified set for Automatic=TRUE) of those values?

Reply:
Hrmm… We will be the latter, and I wonder whether we actually should use NVRAM for this instead of our own more secure interface. We will most likely still have to expose these variables to the operating system though, as it needs them to generate personalised manifests.